src/response.rs

```diff
@@ -210,11 +264,32 @@ impl RecordKind {
                 port,
                 target: target.to_string(),
             },
-            RData::TXT(ref txt) => RecordKind::TXT(
-                txt.iter()
-                    .map(|bytes| String::from_utf8_lossy(bytes).into_owned())
-                    .collect(),
-            ),
+            RData::TXT(ref txt) => {
+                let mut txt_records: HashMap<String, TxtRecordValue> = HashMap::new();
+                for txt_record in txt.iter() {
+                    let mut kv_split = txt_record.split(|c| c == &b'=');
+                    if let Some(key_bytes) = kv_split.next() {
+                        let key = String::from_utf8_lossy(key_bytes).into_owned();
```

Comment:
We could also just canonicalize it to upper or lowercase at this stage if we don't want to introduce a new type, unless losing casing is loss of information. I would say that it shouldn't be a loss of information based on the RFC though.

Reply:
The lookup after serialization would end up being difficult if we go that route. If we choose to canonicalize to all lower case, how do we handle someone looking up a value with a key that is in pascal case?

[http::header](https://docs.rs/http/latest/http/header/struct.HeaderName.html) might be a good place to to find inspiration.

Reply:
> how do we handle someone looking up a value with a key that is in pascal case?

My personal feelings on that are that the hashing semantics would make it clear to the caller that they would need to have a case match when performing a lookup and the canonical format could be documented in the code. But I also concede that it's maybe not the most ergonomic convention.

I'm not against introducing a thin wrapper type like a `TxtRecordKey(String)` tuple struct. It could even be decorated with the serde transparent for de/ser. The hash implementation could be straight forward as well; we could hash on the all upper or all lower variant perhaps.

Reply:
I agree with the loss of information when canonicalizing here and like the idea of a thin wrapper.

```rust
#[derive(Eq, Clone, Debug, serde::Serialize, serde::Deserialize)]
pub struct TxtRecordKey(String);

impl Hash for TxtRecordKey {
    fn hash<H: Hasher>(&self, state: &mut H) {
        self.0.to_lowercase().hash(state)
    }
}

impl PartialEq<Self> for TxtRecordKey {
    fn eq(&self, other: &Self) -> bool {
        self.0.to_lowercase() == other.0.to_lowercase()
    }
}
```

And we can still present the ergonomic key str instead of a wrapped one through the accessor API
```rust
pub fn txt_records(&self) -> impl Iterator<Item = (&str, &TxtRecordValue)>
```

Reply:
I just pushed up the changes where the `TXT` variant still contains String but the wrapper is only used during parsing. That way, case-insensitive handling is only limited to key duplicates in a hashmap and any consumer should still be able to lookup using the intended case. The case would be the one from the first key but that's fine since any subsequent keys irrespective of the case are to be silently ignored as per the spec.

Reply:
I'm happy with this implementation but I'd like @FreeMasen to weigh in one last time because he usually has very good intuition/insight RE: ergonomics.

Reply:
Pushed up the changes with unicase string key.